building-with-ocean/compute-to-data/user-defined-parameters.md

```diff
@@ -0,0 +1,183 @@
+```
+
+#### Step 2: Publish dataset asset with compute service
+
+The publisher now must provide the file URL as `https://example.com` while publishing the asset, as shown in the below image.
+
+![](../../.gitbook/assets/compute-to-data-parameters-publish-dataset.png)
+
+To view a complete tutorial on publishing asset using Ocean Marketplace click [here](../../using-ocean-market/marketplace-publish-data-asset.md).
+
+### Publish an algorithm that uses custom parameters
+
+#### Use case for algorithm custom parameters
+
+For example, if the algorithm publisher has published an URL `https://example.org` which serves python script to analyze historical weather data published in the previous section. If the algorithm publisher wants buyers to specify the number of iterations the algorithm must perform over the data, it is possible to do so using algorithm custom parameters.
+
+Suppose the algorithm publisher defines a parameter called `iterations` and expects the buyer to give this input before running the algorithm in a Compute-to-Data environment. The buyer can enter the desired parameter value using ocean.py or ocean.js.
+
+The provider passes the entered parameters and saves them in a specific path in the Compute-to-Data environment. The algorithm can later read this value and perform required computations.
+
+The following steps will specify how the algorithm publisher can support additional algorithm custom parameters.
+
+#### Step 1: Create an algorithm
+
+The code snippet is only for demo purposes and not for production use.
+
+```python
+def run_algorithm(i: int):
+    pass
+def read_algorithm_custom_input():
+    parameters_file = os.path.join(os.sep, "data", "inputs", "algoCustomData.json")
+    with open(parameters_file, "r") as file:
+        return json.load(file)
+algorithm_inputs = read_algorithm_custom_input()
+iterations = algorithm_inputs["iterations"]
+for i in range(iterations):
+    # Run some machine learning algorithm
+    print(f"Running iteration {i}")
+    result = run_algorithm(i)
+output_dir = os.path.join(os.sep, "data", "outputs")
+with open(os.path.join(output_dir, "result"), "w") as f:
+    f.write(result)
+```
+
+#### Step 2: Publish algorithm asset
+
+The publisher now must provide the file URL as `https://example.org` while publishing the algorithm asset, as shown in the below image.
+
+![](../../.gitbook/assets/compute-to-data-parameters-publish-algorithm.png)
+
+To view a complete tutorial on publishing asset using Ocean Marketplace click [here](../../using-ocean-market/marketplace-publish-data-asset.md).
+
+### Starting compute job with custom parameters
+
+After the publishing, assume that a buyer wants to run the with certain parameters algorithm on selected dataset. The code snippet below shows how the buyer can start the compute job with custom parameter values. Searching dataset, allowing algorithm to run the dataset, buying datatokens, setting up ocean.py/ocean.js is out of scope of this tutorial.
```

Comment:
```suggestion
In this example, the buyer wants to run the algorithm with certain parameters on a selected dataset. The code snippet below shows how the buyer can start the compute job with custom parameter values. Searching the dataset, allowing the algorithm to run on the dataset, buying datatokens, setting up ocean.py/ocean.js is out of scope of this tutorial.
```

Reply:
Rather than listing the things that are out of scope, it would be better to provide a list of links where they can read about how to do those things. 

e.g. "Before embarking on this tutorial you should familiarise yourself with how to [setup ocean.js/ocean.py](link), [buy datatokens](link), [run algorithms](link) etc..."